### PR TITLE
feat(core): Parent workflows should wait for sub-workflows to finish

### PR DESCRIPTION
**Pull request by netroy**
## Summary

This PR
1. Puts parent workflows into waiting state when a sub-workfow is put to wait.
2. Resumes any parent workflows when a resumed sub-workflow finishes

## Related Linear tickets, Github issues, and Community forum posts

CAT-337

## Review / Merge checklist

- [x] PR title and summary are descriptive. ([conventions](../blob/master/.github/pull_request_title_conventions.md)) 
- [ ] [Docs updated](https://github.com/n8n-io/n8n-docs) or follow-up ticket created.
- [x] Tests included. 
- [ ] PR Labeled with `release/backport` (if the PR is an urgent fix that needs to be backported)

### CODE REVIEW

File: packages/cli/src/__tests__/wait-tracker.test.ts [hunk 1]
```diff
@@ -137,6 +148,65 @@ describe('WaitTracker', () => {
...
+					workflowData: execution.workflowData,
+					projectId: project.id,
+					pushRef: execution.data.pushRef,
+				},
+				false,
+				false,
+				execution.id,
+			);
+
+			postExecutePromise.resolve(mock<IRun>());
```
Comment (netroy) [inline packages/cli/src/__tests__/wait-tracker.test.ts:192]: I can't get this promise to resolve in time. It's always resolving after the test ends. any ideas?

**Review by ivov — COMMENTED** [review]
Nice! Looks good, I'll test it with fresh eyes on Monday.

File: packages/core/src/node-execution-context/base-execute-context.ts [hunk 2]
```diff
@@ -97,6 +97,13 @@ export class BaseExecuteContext extends NodeExecutionContext {
 		);
 	}
 
+	async putExecutionToWait(waitTill: Date): Promise<void> {
+		this.runExecutionData.waitTill = waitTill;
+		if (this.additionalData.setExecutionStatus) {
+			this.additionalData.setExecutionStatus('waiting');
+		}
```
[the remark this answers is not in the document]
Comment (netroy) [reply packages/core/src/node-execution-context/base-execute-context.ts:104]: no idea. I removed the `async`, but then didn't want to change the `IExecuteFunctions` interface in a backward incompatible way, so reverted it.
Maybe another thing for v2?

[the remark this answers is not in the document]
Comment (ivov) [reply packages/core/src/node-execution-context/base-execute-context.ts:104]: Tracking: https://linear.app/n8n/issue/CAT-389/make-setexecutionstatus-sync

**Review by ivov — DISMISSED** [review]
Working nicely, great job! 🎉

File: packages/core/src/node-execution-context/base-execute-context.ts [hunk 3]
```diff
@@ -122,9 +127,11 @@ export class BaseExecuteContext extends NodeExecutionContext {
 			parentCallbackManager,
 		});
 
-		// If a sub-workflow execution goes into the waiting state, then put the parent workflow execution also into the waiting state
+		// If a sub-workflow execution goes into the waiting state
 		if (result.waitTill) {
-			await this.putExecutionToWait(result.waitTill);
+			// then put the parent workflow execution also into the waiting state,
+			// but do not use the sub-workflow `waitTill` to avoid WaitTracker resuming the parent execution at the same time as the sub-workflow
+			await this.putExecutionToWait(new Date(WAIT_TIME_UNLIMITED));
```
Comment (ivov) [inline packages/core/src/node-execution-context/base-execute-context.ts:134]: I remember this is what you mentioned back at the office. Maybe we should have a `putExecutionToWaitIndefinitely`

**Review by ivov — DISMISSED** [review]
Works great, no longer induces the double-resuming error 🙏🏻

File: packages/core/src/node-execution-context/base-execute-context.ts (hunk 3 above)
Comment (ivov) [reply packages/core/src/node-execution-context/base-execute-context.ts:134]: I'd rather keep this explicit, without defaults. Another option: `putExecutionToWait({ indefinite: true })`

**Review by ivov — APPROVED** [review]
🚀

**Comment by janober** [comment]
Got released with [`n8n@1.72.0`](https://github.com/n8n-io/n8n/releases/tag/n8n%401.72.0)

**Pull request merged**